SmartDeviceLink/SDLTextAndGraphicUpdateOperation.m

```diff
@@ -0,0 +1,493 @@
+//
+//  SDLTextAndGraphicUpdateOperation.m
+//  SmartDeviceLink
+//
+//  Created by Joel Fischer on 8/13/20.
+//  Copyright © 2020 smartdevicelink. All rights reserved.
+//
+
+#import "SDLTextAndGraphicUpdateOperation.h"
+
+#import "SDLArtwork.h"
+#import "SDLConnectionManagerType.h"
+#import "SDLError.h"
+#import "SDLFileManager.h"
+#import "SDLGlobals.h"
+#import "SDLImage.h"
+#import "SDLLogMacros.h"
+#import "SDLMetadataTags.h"
+#import "SDLShow.h"
+#import "SDLTextAndGraphicState.h"
+#import "SDLVersion.h"
+#import "SDLWindowCapability.h"
+#import "SDLWindowCapability+ScreenManagerExtensions.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface SDLTextAndGraphicUpdateOperation ()
+
+@property (weak, nonatomic) id<SDLConnectionManagerType> connectionManager;
+@property (weak, nonatomic) SDLFileManager *fileManager;
+@property (strong, nonatomic) SDLWindowCapability *currentCapabilities;
+@property (strong, nonatomic) SDLTextAndGraphicState *updatedState;
+
+@property (copy, nonatomic, nullable) CurrentDataUpdatedHandler currentDataUpdatedHandler;
+@property (copy, nonatomic, nullable) SDLTextAndGraphicUpdateCompletionHandler updateCompletionHandler;
+
+@property (copy, nonatomic, nullable) NSError *internalError;
+
+@end
+
+@implementation SDLTextAndGraphicUpdateOperation
+
+- (instancetype)initWithConnectionManager:(id<SDLConnectionManagerType>)connectionManager fileManager:(SDLFileManager *)fileManager currentCapabilities:(SDLWindowCapability *)currentCapabilities currentScreenData:(SDLShow *)currentData newState:(nonnull SDLTextAndGraphicState *)newState currentScreenDataUpdatedHandler:(nullable CurrentDataUpdatedHandler)currentDataUpdatedHandler updateCompletionHandler:(nullable SDLTextAndGraphicUpdateCompletionHandler)updateCompletionHandler {
+    self = [self init];
+    if (!self) { return nil; }
+
+    _connectionManager = connectionManager;
+    _fileManager = fileManager;
+    _currentCapabilities = currentCapabilities;
+    _currentScreenData = currentData;
+    _updatedState = newState;
+    _currentDataUpdatedHandler = currentDataUpdatedHandler;
+    _updateCompletionHandler = updateCompletionHandler;
+
+    return self;
+}
+
+- (void)start {
+    [super start];
+    if (self.cancelled) {
+        // Make sure the update handler is called
+        self.internalError = [NSError sdl_textAndGraphicManager_pendingUpdateSuperseded];
+        [self finishOperation];
+        return;
+    }
+
+    // Build a show with everything from `self.newState`, we'll pull things out later if we can.
+    SDLShow *fullShow = [[SDLShow alloc] init];
+    fullShow.alignment = self.updatedState.alignment;
+    fullShow = [self sdl_assembleShowText:fullShow];
+    fullShow = [self sdl_assembleShowImages:fullShow];
+
+    __weak typeof(self) weakSelf = self;
```

Comment:
If I change the current template and the display capabilities change, then a `Show` request is unnecessary sent. 

For example:
1. Set the template to `NON-MEDIA`.
1. Set the four text-fields: 
```
{
    alignment = CENTERED;
    mainField1 = "textfield 1";
    mainField2 = "textfield 2";
    mainField3 = "textfield 3";
    mainField4 = "textfield 4";
    mediaTrack = "";
    metadataTags = "{\n    mainField1 =     (\n    );\n    mainField2 =     (\n    );\n    mainField3 =     (\n    );\n    mainField4 =     (\n    );\n}";
    templateTitle = "";
}
```
3. Set the template to `DOUBLE_GRAPHIC_WITH_SOFTBUTTONS`. Since the template updated, an automatic update occurs in the `SDLTextAndGraphicManager`'s `sdl_displayCapabilityDidUpdate`. I would expect that a `Show` request would not get sent because no information has changed. However, a `Show` request gets sent resetting all the textfields:
```
{
    alignment = CENTERED;
    mainField1 = "";
    mainField2 = "";
    mainField3 = "";
    mainField4 = "";
    mediaTrack = "";
    metadataTags = "{\n}";
    templateTitle = "";
}
```

Reply:
In addition, if I set the template to `DOUBLE_GRAPHIC_WITH_SOFTBUTTONS` and then try to set `textField1` to a string, then I would expect that a `Show` request would not be sent because the template does not support textFields). However, this `Show` does get sent:
```
{
    alignment = CENTERED;
    mainField1 = "";
    mainField2 = "";
    mainField3 = "";
    mainField4 = "";
    mediaTrack = "";
    metadataTags = "{\n}";
    templateTitle = "";
}
```

Reply:
In regards to the first issue you note, this is correct behavior. Think about the reverse, going from double graphic to NON_MEDIA. If the developer set several text fields (which would then not be shown), then the user switched to the `NON_MEDIA` template, those text fields should show up. But they won't if the change you requested is made.

Reply:
Regarding the second issue, I think that this is legitimate, but I do worry about the complexity of the fix. Basically we would need to check if the developer set data in the `currentState` but the show that will be sent is the same as the current screen data. That will actually get pretty complicated and is fairly error prone. I think it's better to send the few bytes of data in this case rather than to try to do all that calculation.

We can't just check if it's blank, because that only works in some cases, and because sometimes a blank Show is intended to clear the screen.